libazureinit/src/provision/mod.rs

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+pub mod hostname;
+pub mod password;
+pub mod user;
+
+use std::{
+    fs::Permissions,
+    io::Write,
+    marker::PhantomData,
+    os::unix::fs::{DirBuilderExt, PermissionsExt},
+};
+
+use private::*;
+use tracing::instrument;
+
+use crate::{error::Error, imds::PublicKeys};
+
+// State transitions for the Provision structure which are used to force
+// the user to provide backends for things they wish to provision.
+mod private {
+    #[derive(Default)]
+    pub struct NoBackends;
+    #[derive(Default)]
+    pub struct HostnameBackend;
+    #[derive(Default)]
+    pub struct PasswordBackend;
+    #[derive(Default)]
+    pub struct AllBackends;
+}
+
+#[derive(Default)]
+pub struct Provision<State = NoBackends> {
+    hostname: String,
+    username: String,
+    keys: Vec<PublicKeys>,
+    password: Option<String>,
+    hostname_backends: Vec<hostname::Provisioner>,
+    user_backends: Vec<user::Provisioner>,
+    password_backends: Vec<password::Provisioner>,
+    state: PhantomData<State>,
+}
+
+impl Provision<NoBackends> {
+    pub fn new(hostname: String, username: String) -> Self {
+        Self {
+            hostname,
+            username,
+            ..Default::default()
+        }
+    }
+
+    pub fn hostname_provisioners(
```

Comment:
Should `hostname_provisioners` be located under `impl Provision<HostnameBackend>`?